File: src/wings/agent/handoff.py
"""Model handoff detection — when a model returns after other models.

In a main conversation, the candidate pool may select different models
for different turns. When model A is used, then model B, then model A
again, the system injects a handoff prompt so model A knows what
happened in between and can review for needed corrections.
"""

from __future__ import annotations

from datetime import datetime, timezone

from pydantic import BaseModel, Field


class TurnRecord(BaseModel):
    """Record of a single model call within a conversation."""

    turn_id: int
    model_id: str  # API id used, e.g. "anthropic/claude-opus-4-6"
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    user_input_summary: str = ""  # One-line summary of user input
    tool_calls: list[str] = Field(default_factory=list)  # Tool names called
    summary: str = ""  # One-line summary of what the turn accomplished


class HandoffDetector:
    """Detect model handoff and generate context prompts.

    Maintained by AgentLoop. Checks whether the same model is being
    called again after other models intervened.
    """

    def detect(
        self,
        current_model: str,
        turn_history: list[TurnRecord],
    ) -> str | None:
        """Detect if a handoff prompt is needed.

        Returns a handoff prompt string, or None if not needed.

        Trigger: model A was used → other models were used in between
        → model A is being used again.
        """
        if len(turn_history) < 2:
            return None

        # Find the last turn where the current model was used
        previous_turn: TurnRecord | None = None
        for turn in reversed(turn_history):
            if turn.model_id == current_model:
                previous_turn = turn
                break

        if previous_turn is None:
            return None  # This model's first appearance — no handoff needed

        # Find intervening turns by other models.
        # Walk backwards from the end. Turns between now and the previous
        # self-turn are "intervening" if they used a different model.
        intermediate: list[TurnRecord] = []
        for turn in reversed(turn_history):
            if turn is previous_turn:
                break  # reached the last self-turn, stop
            if turn.model_id != current_model:
                intermediate.append(turn)

        if not intermediate:
            return None  # No other models in between

        return self._build_prompt(current_model, previous_turn, intermediate)

    def _build_prompt(
        self,
        current_model: str,
        previous_turn: TurnRecord,
        intermediate_turns: list[TurnRecord],
    ) -> str:
        other_models = sorted({t.model_id for t in intermediate_turns})
        turns_desc = "\n".join(
            f"  - [{t.model_id}] {t.summary or t.user_input_summary}"
            for t in reversed(intermediate_turns)
        )

        return (
            f"[System notice] Since your last turn in this conversation "
            f"(turn #{previous_turn.turn_id}), "
            f"{len(intermediate_turns)} turn(s) were handled by other "
            f"models: {', '.join(other_models)}.\n\n"
            f"Work done in between:\n{turns_desc}\n\n"
            f"Before addressing the current task, please:\n"
            f"1. Review these intermediate turns for issues that need "
            f"correction but haven't been addressed (e.g. inconsistent "
            f"code style, conflicting decisions, missed edge cases)\n"
            f"2. If issues are found, correct them before proceeding\n"
            f"3. If no corrections are needed, proceed with the current task"
        )

File: src/agent/handoff.ts
/**
 * Model handoff detection — when a model returns after other models intervened.
 *
 * In a main conversation, the candidate pool may select different models for
 * different turns. When model A is used, then model B, then model A again,
 * the system injects a handoff prompt so model A knows what happened in
 * between and can review for needed corrections.
 */

export interface TurnRecord {
  turn_id: number;
  /** Internal pool key, e.g. "anthropic/claude-opus-4-6". */
  model_id: string;
  /** Provider name, e.g. "anthropic". */
  provider_name: string;
  /** API model name, e.g. "claude-opus-4-6". */
  service_model: string;
  timestamp: Date;
  /** One-line summary of user input. */
  user_input_summary: string;
  /** Tool names called during this turn. */
  tool_calls: string[];
  /** One-line summary of what the turn accomplished. */
  summary: string;
}

export function makeTurnRecord(
  turn_id: number,
  model_id: string,
  opts: Partial<TurnRecord> = {},
): TurnRecord {
  return {
    turn_id,
    model_id,
    provider_name: opts.provider_name ?? "",
    service_model: opts.service_model ?? "",
    timestamp: new Date(),
    user_input_summary: opts.user_input_summary ?? "",
    tool_calls: opts.tool_calls ?? [],
    summary: opts.summary ?? "",
  };
}

export class HandoffDetector {
  /**
   * Detect if a handoff prompt is needed.
   *
   * Returns a handoff prompt string, or null if not needed.
   *
   * Trigger: model A was used → other models were used in between
   * → model A is being used again.
   */
  detect(currentModel: string, turnHistory: TurnRecord[]): string | null {
    if (turnHistory.length < 2) return null;

    // Find the last turn where the current model was used.
    let previousTurn: TurnRecord | null = null;
    for (let i = turnHistory.length - 1; i >= 0; i--) {
      if (turnHistory[i]!.model_id === currentModel) {
        previousTurn = turnHistory[i]!;
        break;
      }
    }
    if (!previousTurn) return null; // first appearance

    // Find intervening turns by other models.
    const intermediate: TurnRecord[] = [];
    for (let i = turnHistory.length - 1; i >= 0; i--) {
      const t = turnHistory[i]!;
      if (t === previousTurn) break;
      if (t.model_id !== currentModel) intermediate.push(t);
    }

    if (intermediate.length === 0) return null;
    return this._buildPrompt(currentModel, previousTurn, intermediate);
  }

  private _buildPrompt(
    currentModel: string,
    previousTurn: TurnRecord,
    intermediateTurns: TurnRecord[],
  ): string {
    const otherModels = [
      ...new Set(intermediateTurns.map((t) => t.service_model || t.model_id)),
    ].sort();
    const turnsDesc = intermediateTurns
      .reverse()
      .map(
        (t) =>
          `  - [${t.service_model || t.model_id}] ${t.summary || t.user_input_summary}`,
      )
      .join("\n");

    return (
      `[System notice] Since your last turn in this conversation ` +
      `(turn #${previousTurn.turn_id}) as ${currentModel}, ` +
      `${intermediateTurns.length} turn(s) were handled by other ` +
      `models: ${otherModels.join(", ")}.\n\n` +
      `Work done in between:\n${turnsDesc}\n\n` +
      `Before addressing the current task, please:\n` +
      `1. Review these intermediate turns for issues that need ` +
      `correction but haven't been addressed (e.g. inconsistent ` +
      `code style, conflicting decisions, missed edge cases)\n` +
      `2. If issues are found, correct them before proceeding\n` +
      `3. If no corrections are needed, proceed with the current task`
    );
  }
}
